mqtt/mqtt-broker/src/persist.rs

```diff
@@ -1,28 +1,385 @@
     type Error: Into<Error>;
 
-    async fn load(&mut self) -> Result<BrokerState, Self::Error>;
+    async fn load(&mut self) -> Result<Option<BrokerState>, Self::Error>;
 
     async fn store(&mut self, state: BrokerState) -> Result<(), Self::Error>;
 }
 
+/// A persistor that does nothing.
+#[derive(Debug)]
 pub struct NullPersistor;
 
 #[async_trait]
 impl Persist for NullPersistor {
     type Error = Error;
 
-    async fn load(&mut self) -> Result<BrokerState, Self::Error> {
-        Ok(BrokerState::default())
+    async fn load(&mut self) -> Result<Option<BrokerState>, Self::Error> {
+        Ok(None)
     }
 
     async fn store(&mut self, _: BrokerState) -> Result<(), Self::Error> {
         Ok(())
     }
 }
+
+/// An abstraction over the broker state's file format.
+pub trait FileFormat {
+    type Error: Into<Error>;
+
+    /// Load `BrokerState` from a reader.
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error>;
+
+    /// Store `BrokerState` to a writer.
+    fn store<W: Write>(&self, writer: W, state: BrokerState) -> Result<(), Self::Error>;
+}
+
+/// Loads/stores the broker state to a file defined by `FileFormat`.
+///
+/// The `FilePersistor` works by storing to a new state file and then updating
+/// a symlink to `state.dat`.
+///
+/// Loading always reads from `state.dat`. This allows for saving multiple
+/// copies of the state and rollback in case something gets corrupted.
+/// It also aids in "transactionally" saving the state. Updating the symlink
+/// "commits" the changes. This is an attempt to prevent corrupting the state
+/// file in case the process crashes in the middle of writing.
+#[derive(Debug)]
+pub struct FilePersistor<F> {
+    dir: PathBuf,
+    format: F,
+    previous_count: usize,
+
+    /// seq is bumped on every store to disambiguate stores in the same timestamp
+    seq: u16,
+}
+
+impl<F> FilePersistor<F> {
+    pub fn new<P: Into<PathBuf>>(dir: P, format: F) -> Self {
+        FilePersistor {
+            dir: dir.into(),
+            format,
+            previous_count: STATE_DEFAULT_PREVIOUS_COUNT,
+            seq: 0,
+        }
+    }
+
+    /// Sets the number of past states to save
+    ///
+    /// The intent is to allow rollback to a pervious state.
+    /// The default is `2`, which saves the current and previous state.
+    /// The minimum value is `1` (the current state).
+    pub fn with_previous_count(mut self, previous_count: usize) -> Self {
+        self.previous_count = cmp::max(1, previous_count);
+        self
+    }
+}
+
+/// A simple format based on Bincode and serde
+#[derive(Clone, Debug, Default)]
+pub struct BincodeFormat;
+
+impl BincodeFormat {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl FileFormat for BincodeFormat {
+    type Error = Error;
+
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error> {
+        let decoder = GzDecoder::new(reader);
+        fail_point!("bincodeformat.load.deserialize_from", |_| {
```

Comment:
I wonder if we can make fail_point! to be aware of our error kinds in a way, that we can use it like this
```rust
fail_point_ext(ErrorKind::Persist(ErrorReason::Deserialize));
```
that generates the same output. 
It may require `ErrorKind::Persist` to implement some custom trait to return the correct string though. Probably with proc_macro :)

Reply:
i had considered that. perhaps we can optimize this in a future PR?